src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -125,29 +125,48 @@ private void DetermineSerializationCapabilities(JsonIgnoreCondition? ignoreCondi
             }
         }
 
-        private void DetermineIgnoreCondition(JsonIgnoreCondition? ignoreCondition, bool isReferenceType)
+        private void DetermineIgnoreCondition(JsonIgnoreCondition? ignoreCondition, bool defaultValueIsNull)
         {
             if (ignoreCondition != null)
             {
                 Debug.Assert(MemberInfo != null);
                 Debug.Assert(ignoreCondition != JsonIgnoreCondition.Always);
 
-                if (ignoreCondition != JsonIgnoreCondition.Never)
+                if (ignoreCondition == JsonIgnoreCondition.WhenWritingDefault)
                 {
-                    Debug.Assert(ignoreCondition == JsonIgnoreCondition.WhenWritingDefault);
                     IgnoreDefaultValuesOnWrite = true;
                 }
+                else if (ignoreCondition == JsonIgnoreCondition.WhenWritingNull)
+                {
+                    if (defaultValueIsNull)
+                    {
+                        IgnoreDefaultValuesOnWrite = true;
+                    }
+                    else
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_IgnoreConditionOnValueTypeInvalid(this);
+                    }
+                }
+                // Else JsonIgnoreCondition == JsonIgnoreCondition.Never
```

Comment:
Remove comment?

Reply:
Will remove - fwiw, I intended for it to be a pseudo `Debug.Assert`, since I can't add an `else` branch to do it.